# Retail sales insights and trend analysis

In [1]:
#Import dependencies

import pandas as pd
import numpy as np

#Importing postgre toolkit

import psycopg2 # this works as an adapter
from sqlalchemy import create_engine


In [2]:
!pip install psycopg2-binary sqlalchemy


In [3]:
#Load the dataset

df = pd.read_csv(r"C:\Bindhuj`s OneDrive\OneDrive\Desktop\New folder\DATA SCIENCE\PROJECTS\WALMART SALES ANALYSIS\Walmart.csv")

In [4]:
#Shape of dataset

df.shape

(10051, 11)

* The dataset have 10051 rows and 11 columns

In [5]:
#View the dataset

df.head()

invoice_id   Branch         City                category unit_price  \
0           1  WALM003  San Antonio       Health and beauty     $74.69   
1           2  WALM048    Harlingen  Electronic accessories     $15.28   
2           3  WALM067  Haltom City      Home and lifestyle     $46.33   
3           4  WALM064      Bedford       Health and beauty     $58.22   
4           5  WALM013       Irving       Sports and travel     $86.31   

   quantity      date      time payment_method  rating  profit_margin  
0       7.0  05/01/19  13:08:00        Ewallet     9.1           0.48  
1       5.0  08/03/19  10:29:00           Cash     9.6           0.48  
2       7.0  03/03/19  13:23:00    Credit card     7.4           0.33  
3       8.0  27/01/19  20:33:00        Ewallet     8.4           0.33  
4       7.0  08/02/19  10:37:00        Ewallet     5.3           0.48

In [6]:
# Inforrmation of dataset

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10051 entries, 0 to 10050
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   invoice_id      10051 non-null  int64  
 1   Branch          10051 non-null  object 
 2   City            10051 non-null  object 
 3   category        10051 non-null  object 
 4   unit_price      10020 non-null  object 
 5   quantity        10020 non-null  float64
 6   date            10051 non-null  object 
 7   time            10051 non-null  object 
 8   payment_method  10051 non-null  object 
 9   rating          10051 non-null  float64
 10  profit_margin   10051 non-null  float64
dtypes: float64(3), int64(1), object(7)
memory usage: 863.9+ KB


* Unit price need to be integer
* There is no null values

In [7]:
# Check for duplicates

df.duplicated().sum()

51

In [8]:
# Drop duplicated

df.drop_duplicates(inplace = True)

* There are total 51 duplicates in the dataset

In [9]:
# Check for missing values
df.isnull().sum()

invoice_id         0
Branch             0
City               0
category           0
unit_price        31
quantity          31
date               0
time               0
payment_method     0
rating             0
profit_margin      0
dtype: int64

* Unit_price and quantity has 31 missing values

In [10]:
# Drop records with missing values

df.dropna(inplace = True)

In [11]:
# Check shape of dataset

df.shape

(9969, 11)

* There are total 9969 rows and 11 columns in the dataset

In [12]:
# Change datatype of unit price

df['unit_price'].astype(int)

ValueError: invalid literal for int() with base 10: '$74.69'

In [14]:
# Replace $ from unit price and convert to float datatype

df['unit_price'] = df['unit_price'].str.replace('$', '').astype(float)
df.head()

invoice_id   Branch         City                category  unit_price  \
0           1  WALM003  San Antonio       Health and beauty       74.69   
1           2  WALM048    Harlingen  Electronic accessories       15.28   
2           3  WALM067  Haltom City      Home and lifestyle       46.33   
3           4  WALM064      Bedford       Health and beauty       58.22   
4           5  WALM013       Irving       Sports and travel       86.31   

   quantity      date      time payment_method  rating  profit_margin  
0       7.0  05/01/19  13:08:00        Ewallet     9.1           0.48  
1       5.0  08/03/19  10:29:00           Cash     9.6           0.48  
2       7.0  03/03/19  13:23:00    Credit card     7.4           0.33  
3       8.0  27/01/19  20:33:00        Ewallet     8.4           0.33  
4       7.0  08/02/19  10:37:00        Ewallet     5.3           0.48

In [15]:
# Returns all the columns

df.columns

Index(['invoice_id', 'Branch', 'City', 'category', 'unit_price', 'quantity',
       'date', 'time', 'payment_method', 'rating', 'profit_margin'],
      dtype='object')

In [16]:
df.columns = df.columns.str.lower()

In [17]:
# Feature engineering

df['total'] = df['unit_price'] * df['quantity']
df.head(10)

invoice_id   branch         city                category  unit_price  \
0           1  WALM003  San Antonio       Health and beauty       74.69   
1           2  WALM048    Harlingen  Electronic accessories       15.28   
2           3  WALM067  Haltom City      Home and lifestyle       46.33   
3           4  WALM064      Bedford       Health and beauty       58.22   
4           5  WALM013       Irving       Sports and travel       86.31   
5           6  WALM026       Denton  Electronic accessories       85.39   
6           7  WALM088     Cleburne  Electronic accessories       68.84   
7           8  WALM100       Canyon      Home and lifestyle       73.56   
8           9  WALM066    Grapevine       Health and beauty       36.26   
9          10  WALM065   Texas City      Food and beverages       54.84   

   quantity      date      time payment_method  rating  profit_margin   total  
0       7.0  05/01/19  13:08:00        Ewallet     9.1           0.48  522.83  
1       5.0  08/03/19  10:29:00           Cash     9.6           0.48   76.40  
2       7.0  03/03/19  13:23:00    Credit card     7.4           0.33  324.31  
3       8.0  27/01/19  20:33:00        Ewallet     8.4           0.33  465.76  
4       7.0  08/02/19  10:37:00        Ewallet     5.3           0.48  604.17  
5       7.0  25/03/19  18:30:00        Ewallet     4.1           0.48  597.73  
6       6.0  25/02/19  14:36:00        Ewallet     5.8           0.33  413.04  
7      10.0  24/02/19  11:38:00        Ewallet     8.0           0.18  735.60  
8       2.0  10/01/19  17:15:00    Credit card     7.2           0.33   72.52  
9       3.0  20/02/19  13:27:00    Credit card     5.9           0.33  164.52

In [ ]:
help(create_engine)

In [18]:
#Connecting to postgresql
#"postgresql+psycopg2://user:password@localhost:3306/db_name"
engine_psql = create_engine("postgresql+psycopg2://postgres:@localhost:5432/Sales_db")
try:
    engine_psql
    print("Connection Successed to PSQL")
except:
    print("Unable to connect")

Connection Successed to PSQL


In [19]:
df.to_sql(name='retail_sales', con=engine_psql, if_exists='append', index=False)

969

In [20]:
df.to_csv('retail_clean_data.csv', index=False)